### PR TITLE
Ampers: Zheng Cao & Nicoleta Brandolini

### PR DESCRIPTION
**Pull request by nbrandolini**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the advantages of using git when collaboratively working on one code base? |It allows all members to share their codes by pushing to repo and access updated version of code base by pulling from repo. It is a very efficient way to collaborate remotely, which allows team members to work on problems individually on different computers and share their work after.|
Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? |It’s only tied to the class and not to any particular instance, so it made sense to make it a class method. Any player instance should be able to use it,  without having to create a new score instance every time they play a word.|
Describe an example of a test you wrote for a _nominal case_. |For TileBag#draw_tiles, “it ‘Returns a collection of random tiles’ ” is a nominal case. The data type, length, and element of the returned array was checked.|
Describe an example of a test you wrote for an _edge case_. |For TileBag#draw_tiles, “it ‘Raises an ArgumentError if user inputs an invalid number’ ” is an edge case. Several invalid inputs (string, negative num, too large number) were tested. If the player draws more tiles than there are in the bag, the program will raise an Argument Error and also if the player tries to enter a negative number or not an integer.|
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? |Yes, we used ‘reduce’ in Player#total_score to calculate the total score of words played. With this method, we can implement total_score with only one line of code. |
What was one method you and your pair used to debug code? |For most of bugs we encountered, we were able to find the cause by reading Error message. If this didn’t work, we would use “binding.pry” to debug.|
Is there a specific piece of code you'd like feedback on? |1) Scoring.highest_score_from looks cumbersome to us. We would like to see a better way to implement this method. 2) Also, if we can get some feedback on the  edge-cases we implemented, it will be helpful.|
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? |1) NB - Working with Zheng was very helpful for me, as I learned from her way of structuring the project first and then take it step-by-step. While she guided me through the project she was very patient. Her knowledge of the material is way above my knowledge at the moment. We took turns typing and working on different computers. 2) ZC - Sometimes, we got confused about nominal cases and edge cases for a method. The fact that we were able to utilize most of the new techniques (ternary conditional, post-fix conditional, git push and pull, debugging with Stack Traces, CONSTANTS, Enumerable’s built-in methods, and handling Exceptions) in our project made us feel really proud. :)|

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,6 @@
 *.gem
 *.rbc
+.vscode/
```
Comment (CheezItMan) [inline .gitignore:3]: Interesting, you're using VS Code

File: lib/scoring.rb [hunk 2]
```diff
@@ -1,9 +1,55 @@
+require "pry"
+
 module Scrabble
   class Scoring
+
+    SCORE_CHART = {
```
Comment (CheezItMan) [inline lib/scoring.rb:6]: Good way to store the letter scores, and nice use of a CONSTANT.

File: lib/scoring.rb [hunk 3]
```diff
@@ -1,9 +1,55 @@
...
+
+    # Returns total score for the given word
+    # Seven letter words receive 50 bonus points
     def self.score(word)
+      return nil if word.length > 7 || word.length == 0
+
+      score = 0
+      word.each_char do |c|
+        return nil if !SCORE_CHART.keys.include?(c.upcase)
+        SCORE_CHART.each do |key, value|
```
Comment (CheezItMan) [inline lib/scoring.rb:23]: You actually don't need this loop, you can instead just do:
```ruby
score += SCORE_CHART[c.upcase]
```

Take advantage of the fact that this is a hash.

File: specs/scoring_spec.rb [hunk 4]
```diff
@@ -42,21 +43,36 @@
...
     it 'returns the only word in a length-1 array' do
+      Scrabble::Scoring.highest_score_from(["word"]).must_equal "word"
     end
 
     it 'returns the highest word if there are two words' do
+      Scrabble::Scoring.highest_score_from(["Nicole", "Zheng"]).must_equal "Zheng"
     end
 
     it 'if tied, prefer a word with 7 letters' do
+      Scrabble::Scoring.highest_score_from(["ZZZZZZ", "AAAAADB"]).must_equal "AAAAADB"
```
Comment (CheezItMan) [inline specs/scoring_spec.rb:58]: Nice, you considered the order!

File: lib/player.rb [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    end
+
+    # Returns the highest scoring played word
+    def highest_scoring_word
+      return Scrabble::Scoring.highest_score_from(@plays)
+    end
+
+    # Returns the `highest_scoring_word` score
+    def highest_word_score
+      return Scrabble::Scoring.score(highest_scoring_word)
```
Comment (CheezItMan) [inline lib/player.rb:52]: Well done

File: lib/player.rb [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    end
+
+    # A collection of letters that the player can play (max 7)
+    def tiles
+      return @tiles
+    end
+
+    # Draw 7 tiles from the given tile bag
+    def draw_tiles(tile_bag)
+      tile_bag = Scrabble::TileBag.new
```
Comment (CheezItMan) [inline lib/player.rb:62]: You have been sent a `TileBag` instance as a parameter just use that you should **NOT** create a new instance of `TileBag`!

File: specs/player_spec.rb [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+require_relative '../lib/player'
+require_relative '../lib/scoring'
+require_relative '../lib/tile_bag'
+require 'pry'
+
+Minitest::Reporters.use! Minitest::Reporters::SpecReporter.new
+
+describe 'Player' do
+  before do
+    @player = Scrabble::Player.new("Paul")
```
Comment (CheezItMan) [inline specs/player_spec.rb:14]: A `before` block!  Nicely done

File: specs/player_spec.rb [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    end
+
+    it "Returns ArgumentError if the word is already present" do
+      @player.play("banana")
+
+      proc {@player.play("banana")}.must_raise ArgumentError
+    end
+  end
+
+  describe 'total_score' do
```
Comment (CheezItMan) [inline specs/player_spec.rb:64]: What happens to the score if they've won and you try to play another word?  The score shouldn't change.  You should **test** that!

File: specs/player_spec.rb [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+    end
+
+    it "Return false if score is under 100 points" do
+      @player.play("aaaaaaa")
+
+      @player.won?.must_equal false
+    end
+  end
+
+  describe 'highest_scoring_word' do
```
Comment (CheezItMan) [inline specs/player_spec.rb:91]: What's the result of this method if nothing has been played.

File: specs/player_spec.rb [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+  describe 'tiles' do
+    it "Returns a collection of letters that the player can play" do
+      @player.tiles.must_be_instance_of Array
+      @player.tiles.length.must_be :<, 8
+    end
+  end
+
+  describe 'draw_tiles' do
+    it "Fills tiles array until it has 7 letters from the given tile bag" do
```
Comment (CheezItMan) [inline specs/player_spec.rb:117]: You should also probably remove some tiles from the player's hand (like 2) and then call `draw_tiles` and see if it draws the right number of tiles (like 5).

File: lib/tile_bag.rb [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    attr_reader :tiles, :number_of_tiles
+
+    LETTER_QUANTITIES = {
+      "A" => 9, "B" => 2, "C" => 2, "D" => 4, "E" => 12, "F" => 2, "G" => 3,
+      "H" => 2, "I" => 9, "J" => 1, "K" => 1, "L" => 4, "M" => 2, "N" => 6,
+      "O" => 8, "P" => 2, "Q" => 1, "R" => 6, "S" => 4, "T" => 6,
+      "U" => 4, "V" => 2, "W" => 2, "X" => 1, "Y" => 2, "Z" => 1 }
+
+    def initialize
+      @tiles = []
```
Comment (CheezItMan) [inline lib/tile_bag.rb:15]: This sets `@tiles` to an empty array and doesn't put any tiles in it.

File: lib/tile_bag.rb [hunk 12]
```diff
@@ -0,0 +1,35 @@
+require 'pry'
+
+module Scrabble
+  class TileBag
+
+    attr_reader :tiles, :number_of_tiles
+
+    LETTER_QUANTITIES = {
```
Comment (CheezItMan) [inline lib/tile_bag.rb:8]: This is great for the letter quantities, but you're using this hash to store the tiles in the bag.  This is problematic for a couple of reasons.
1.  All instances of `TileBag` are **sharing** this constant.  So if you remove a "z" from one `TileBag` instance, you've removed it from them all.
1.  This structure makes any letter  **equally** likely to be drawn from the bag.  However there is only 1 "z" and 12 "E"'s.

File: specs/tile_bag_spec.rb [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+  end
+
+  describe 'initialize' do
+    it "Set-up the instance with a collection of all default tiles" do
+      @tile_bag.must_respond_to :tiles
+      @tile_bag.tiles.must_be_instance_of Array
+    end
+  end
+
+  describe 'draw_tiles' do
```
Comment (CheezItMan) [inline specs/tile_bag_spec.rb:22]: What happens when you draw all the tiles?

Also you need to check that the number of each drawn tile has gone down.

**Comment by CheezItMan** [comment]
# Scrabble
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions |  Check, I hope my solution code helped a bit, there are other solutions as well.
Both Teammates contributed to the codebase | Both team member have commits
Regular Commits with meaningful commit messages |  Good number of commits and good commit messages
**Scoring class** |
`score` method |  Check, but see my in-line note
Uses appropriate data structure to store the letter score |  Check
Appropriately handles edge cases |  Check
`highest_score_from` method |  Check
Appropriately handles edge cases |  Check
Test implementations for `highest_score_from`| Check
**Player class** |
Uses an array to keep track of words played |  Check
Uses existing code to create `highest_scoring_word` and `highest_word_score` | Check
Returns `true` or `false` from `won?` method | Check
Tests edge cases on the `play(word)` | Check, but some issues with `total_score`
**TileBag class** |
Uses a data structure to keep the set of default tiles in the bag |  Check, but see my notes on it.
`draw_tiles` method uses the parameter to draw a set of random tiles | It pick random tiles, but doesn't remove them from the structure.  
`tiles_remaining` method returns a count |  Check
Tests `draw_tiles` edge cases |  No edge-cases tested.
Summary | Not bad, you had some trouble, specifically with `TileBag`. You also need to look a bit more closely for edge-cases.  Let me know where you have questions.
